File: dev/test_data/main/cpp/expected/aas_core_meta.v3/input/snippets/verification/submodel_element_is_of_type.cpp
std::map<
  types::AasSubmodelElements,
  std::function<bool(const std::shared_ptr<types::ISubmodelElement>&)>
> ConstructAasSubmodelElementToIs() {
  std::map<
    types::AasSubmodelElements,
    std::function<bool(const std::shared_ptr<types::ISubmodelElement>&)>
  > result = {
    {
      types::AasSubmodelElements::kAnnotatedRelationshipElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsAnnotatedRelationshipElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kBasicEventElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsBasicEventElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kBlob,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsBlob(*element);
      }
    },
    {
      types::AasSubmodelElements::kCapability,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsCapability(*element);
      }
    },
    {
      types::AasSubmodelElements::kDataElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsDataElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kEntity,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsEntity(*element);
      }
    },
    {
      types::AasSubmodelElements::kEventElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsEventElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kFile,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsFile(*element);
      }
    },
    {
      types::AasSubmodelElements::kMultiLanguageProperty,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsMultiLanguageProperty(*element);
      }
    },
    {
      types::AasSubmodelElements::kOperation,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsOperation(*element);
      }
    },
    {
      types::AasSubmodelElements::kProperty,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsProperty(*element);
      }
    },
    {
      types::AasSubmodelElements::kRange,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsRange(*element);
      }
    },
    {
      types::AasSubmodelElements::kReferenceElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsReferenceElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kRelationshipElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsRelationshipElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kSubmodelElement,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsSubmodelElement(*element);
      }
    },
    {
      types::AasSubmodelElements::kSubmodelElementList,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsSubmodelElementList(*element);
      }
    },
    {
      types::AasSubmodelElements::kSubmodelElementCollection,
      [](const std::shared_ptr<types::ISubmodelElement>& element) {
        return types::IsSubmodelElementCollection(*element);
      }
    }
  };

  #ifdef DEBUG
  for (types::AasSubmodelElements literal : iteration::kOverAasSubmodelElements) {
    const auto it = result.find(literal);
    if (it == result.end()) {
      throw std::logic_error(
        common::Concat(
          "The enumeration literal ",
          std::to_string(static_cast<std::uint32_t>(literal)),
          " of types::AasSubmodelElements "
          " is not covered in ConstructAasSubmodelElementToIs"
        )
      );
    }
  }
  #endif

  return result;
}

const std::map<
  types::AasSubmodelElements,
  std::function<bool(const std::shared_ptr<types::ISubmodelElement>&)>
> kAasSubmodelElementToIs = ConstructAasSubmodelElementToIs();

bool SubmodelElementIsOfType(
  const std::shared_ptr<types::ISubmodelElement>& element,
  types::AasSubmodelElements element_type
) {
  const auto it = kAasSubmodelElementToIs.find(element_type);
  if (it == kAasSubmodelElementToIs.end()) {
    throw std::invalid_argument(
      common::Concat(
        "Unexpected element type: ",
        std::to_string(static_cast<uint32_t>(element_type))
      )
    );
  }

  const std::function<
    bool(const std::shared_ptr<types::ISubmodelElement>&)
  >& is_function(
    it->second
  );

  return is_function(element);
}

File: dev/test_data/main/typescript/expected/aas_core_meta.v3/input/snippets/Verification/submodel_element_is_of_type.ts
const AAS_SUBMODEL_ELEMENTS_TO_IS =
  new Map<AasTypes.AasSubmodelElements, (that: AasTypes.Class) => boolean>(
    [
      [
        AasTypes.AasSubmodelElements.AnnotatedRelationshipElement,
        AasTypes.isAnnotatedRelationshipElement
      ],
      [
        AasTypes.AasSubmodelElements.BasicEventElement,
        AasTypes.isBasicEventElement
      ],
      [
        AasTypes.AasSubmodelElements.Blob,
        AasTypes.isBlob
      ],
      [
        AasTypes.AasSubmodelElements.Capability,
        AasTypes.isCapability
      ],
      [
        AasTypes.AasSubmodelElements.DataElement,
        AasTypes.isDataElement
      ],
      [
        AasTypes.AasSubmodelElements.Entity,
        AasTypes.isEntity
      ],
      [
        AasTypes.AasSubmodelElements.EventElement,
        AasTypes.isEventElement
      ],
      [
        AasTypes.AasSubmodelElements.File,
        AasTypes.isFile
      ],
      [
        AasTypes.AasSubmodelElements.MultiLanguageProperty,
        AasTypes.isMultiLanguageProperty
      ],
      [
        AasTypes.AasSubmodelElements.Operation,
        AasTypes.isOperation
      ],
      [
        AasTypes.AasSubmodelElements.Property,
        AasTypes.isProperty
      ],
      [
        AasTypes.AasSubmodelElements.Range,
        AasTypes.isRange
      ],
      [
        AasTypes.AasSubmodelElements.ReferenceElement,
        AasTypes.isReferenceElement
      ],
      [
        AasTypes.AasSubmodelElements.RelationshipElement,
        AasTypes.isRelationshipElement
      ],
      [
        AasTypes.AasSubmodelElements.SubmodelElement,
        AasTypes.isSubmodelElement
      ],
      [
        AasTypes.AasSubmodelElements.SubmodelElementList,
        AasTypes.isSubmodelElementList
      ],
      [
        AasTypes.AasSubmodelElements.SubmodelElementCollection,
        AasTypes.isSubmodelElementCollection
      ]
    ]);

function assertAllTypesCoveredInAasSubmodelElementsToIs() {
  for (const literal of AasTypes.overAasSubmodelElements()) {
    if (!AAS_SUBMODEL_ELEMENTS_TO_IS.has(literal)) {
      throw new Error(
        `The enumeration literal ${literal} of AasTypes.AasSubmodelElements ` +
          "is not covered in AAS_SUBMODEL_ELEMENTS_TO_IS"
      );
    }
  }
}
assertAllTypesCoveredInAasSubmodelElementsToIs();

/**
 * Check that `element` is an instance of class corresponding to
 * `expectedType`.
 *
 * @param element - to be checked for type
 * @param expectedType - in the check
 * @returns `true` if `element` corresponds to `expectedType`
 */
export function submodelElementIsOfType(
  element: AasTypes.ISubmodelElement,
  expectedType: AasTypes.AasSubmodelElements
): boolean {
  const isFunc = AAS_SUBMODEL_ELEMENTS_TO_IS.get(expectedType);
  return isFunc(element);
}
